keps/sig-node/4800-cpumanager-split-l3cache/README.md

```diff
@@ -0,0 +1,914 @@
+  }
+  ```
+
+- In `static_policy.go`, check for option `AlignByUnCoreCacheOption` in takeByTopology. Implement new cpu assignment function that takeByTopologyUncorecachePacked which adds a third level of CPU grouping in the topology. **Socket->Numa->UnCoreCache**. TakeByTopologyUnCoreCachePacked uses the "packed" cpu sorting strategy similar to takeByTopologyNUMAPacked.
+
+  ```go
+  func (p *staticPolicy) takeByTopology(availableCPUs cpuset.CPUSet, numCPUs int) (cpuset.CPUSet, error) {
+  ...
+    acc.numaOrSocketsFirst.takeFullFirstLevel()
+    if acc.isSatisfied() {
+     return acc.result, nil
+     }
+    acc.numaOrSocketsFirst.takeFullSecondLevel()
+    if acc.isSatisfied() {
+      return acc.result, nil
+    }
+   // 2. Acquire partial uncorecache, if there are enough CPUs available to satisfy the container requirement
+   //    Acquire the full uncorecache, if available and the container requires at least all the CPUs in the uncorecache grouping
+    acc.numaOrSocketsFirst.takeThirdLevel()
+    if acc.isSatisfied() {
+        return acc.result, nil
+    }
+  ...
+  }
+  ```
+
+  - Define a new interface in the `numaOrSocketsFirstFunc` in `cpu_assignment.go` that implements `takeThirdLevel`, which calls `takeUnCoreCache`.
+
+  ```go
+  // In Split L3 Topology, we take from the sets of uncorecache as the third level
+  func (n *numaFirst) takeThirdLevel() {
+    n.acc.takeUnCoreCache()
+  }
+  // First try to take partial uncorecache (CCD), if available and the request size can fit w/in the uncorecache.
+  // Second try to take the full CCD if available and need is at least the size of the uncorecache group.
+  func (a *cpuAccumulator) takeUnCoreCache() {
+    // check if SMT ON
+    
+    for _, uncore := range a.allUnCoreCache() {
+      numCoresNeeded := a.numCPUsNeeded / a.topo.CPUsPerCore() //this is another new change
+      
+      var freeCPUsInUncorecache cpuset.CPUSet
+      //need to get needed cores in uncorecache
+      freeCoresInUncorecache := a.details.CoresNeededInUnCoreCache(numCoresNeeded, uncore)
+      klog.V(2).Info("free cores from a.details list: ", freeCoresInUncorecache)
+      for _, coreId := range freeCoresInUncorecache.List() {
+          freeCPUsInUncorecache = freeCPUsInUncorecache.Union(a.topo.CPUDetails.CPUsInCores(coreId))
+      }
+      klog.V(2).Info("freeCPUsInUncorecache  : ", freeCPUsInUncorecache)
+      if a.numCPUsNeeded == freeCPUsInUncorecache.Size() {
+          klog.V(4).InfoS("takePartialUncore: claiming cores from Uncorecache ID", uncore)
+        a.take(freeCPUsInUncorecache)
+      }
+      //take full Uncorecache if the numCPUsNeeded is greater the L3 cache size
+      a.takeFullUnCore
+      if a.isSatisfied() {
+        return
+      }
+    }
+  }
+  ```
+
+- Implement a new interface in NumaorSocket struct that checks for CPUs in a third level
+
+  ```go
+  type numaOrSocketsFirstFuncs interface {
+  takeFullFirstLevel()
+  takeFullSecondLevel()
+  takeFullThirdLevel()
```

Comment:
mixed feelings about this name. At this point in time third level will always be `L3Groups` with no alternatives on the radar; OTOH we introduced First and Second level because of `align-by-socket`, to support it. I wonder if we need this indirection or if we can just name it `takeL3Groups` or so

Reply:
has there been any options for First and Second? If there hasn't I would say we could just stay consistent with third. I'm not inclined for either option

Reply:
First and Second are redirected internally to handle align-by-socket and depending on the machine HW properties. ATM I don't see this need for L3 grouping, so I'd just avoid a generalization we're not sure about, but I'm not too strongly attached  to this opinion.

Reply:
@ffromani , I actually prefer takeL3Group as opposed to thirdLevel.  I originally wanted to keep it generic in case there are other "uncorecache" groups in future HW arch design but currently I'm not aware of  any trend or roadmaps to support this assumption.  Will rename the func to takeL3Group.

Reply:
I can make the change but I thought we didn't want code in the design details, this comment should be made over there

Reply:
The above comment was made assuming we want or need to keep this part in the KEP. My general recommendation is still to describe the algorithm and highlight the tradeoff and the design decision, adding code snippets where it helps the reader; adding code examples past that is probably wasteful because the code in the KEP gets obsolete very very fast, so the details are better served in the implementation PR.